Update alternatives only after package installation.

ftclausen opened the pull request:
Hello Julian,

I noticed, that on a clean box, the sqlplus recipe was failing by trying to set the alternatives prior to sqlplus package installation. What _I think_ was happening is that it would run the Ruby library code at recipe compile time instead of recipe execution time. This meant that the binaries update-alternatives uses are not present.

This PR is an attempt at fixing this situation by
- Removing the Ruby library 
- Create a ruby_block containing the update alternatives code (that used to be in the library)
- Call this ruby block once the sqlplus RPM has been installed

I modelled this approach on what the Java cookbook does. Perhaps a cleaner way is to figure out how a ruby_block can call a Chef cookbook library. That way the recipe code can be kept tidier however it is still pretty short so I figure it was OK in this instance.

Thanks,

Fred.

juliandunn commented:
Sorry, didn't see this till now. Yes, that looks exactly like what's happening. `install_alternatives` is evaluated and executed at compile time.

I think in the `ruby_block` you can just mix in the helper with a direct `include` statement and that's how you could call the library within the block, if you want to try that. Or I can just merge this as-is since it works.

ftclausen commented:
I actually tried that first but, since the ruby block seems to not be part of a class, it gives a syntax error. 

But, I just thought, maybe it is possible to declare a class in the ruby block so that we can include from there. I'll try that out and let you know.

juliandunn commented:
Hi @ftclausen,

You're right... but you can use some metaprogramming magic to do the include. Example:

``` ruby
ruby_block 'foo' do
  block do
    Chef::Resource::RubyBlock.send(:include, Chef::Mixin::ShellOut)

    cmd = shell_out!('echo hello')
    puts cmd.stdout
  end
end
```

Which gets you something like this:

```
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * ruby_block[foo] action runhello

    - execute the ruby block foo
```

ftclausen commented:
Sorry for my tardy response - that looks promising. I'm testing it out and, if I can get it to work, then I'll probably open a new PR with that version (since it was easier to start again instead of adding the library back here).

juliandunn commented:
Cool. Either works; you can always force-push the branch once you're happy with the changes and all the old diffs will be marked as outdated.

juliandunn commented:
Did you manage to get a chance to look at this @ftclausen?

ftclausen commented:
Sorry about the continued delay - it's been a manic time at work and I've just moved from Europe to Australia. I'll definitely look at it in the coming week though.

ftclausen committed:
Redid sqlplus update-alternatives notification using metaprogramming magic

ftclausen committed:
Merge branch 'master' into package_install_order

Conflicts:
	libraries/utils.rb

Incorporate changes from master "@n" versus "@node"

ftclausen committed:
Remove module "constructor" and initialise @n in ruby block

I was getting nil "@n" objects because the constructor in the
library code was not being called since it was a module, not a class.
Initialising the @n object in the mixed in block instead seems to work.

ftclausen commented:
Again, apologies for my slow response! I have done the following 
- Reset the "package_install_order" back to the starting point with the original library
- Did some initial experimenting and then remembered to merge in your latest changes + resolve conflict.
- Arrived at the seemingly working code you see in the diff

(Note: I am a learning Rubyist) - There were issues with the "n" instance variable being nil due to the "initialize" method not being called upon including the module contained in the library. As I understand it this is due to it being a module and not a class thus "initialize" is not called.

My solution was to remove the (seemingly) spurious initialize method and then, in the mixed in Ruby block, assign the node to "n" instance variable.

It all seems to work in my testing but any feedback welcome (both on the cookbook and Ruby approaches).
